doc/ps_design.md

```diff
@@ -0,0 +1,382 @@
+# Parameter Server Design
+
+
+## Overview
+Currently, there is one parameter server (PS) co-existed with the master. In order to support multiple PSs and PS fault tolerance, we need to separate PS from the master. Besides a KV store for model variables and embedding tables, PS should also support model variable update and embedding table sparse update using gradients.
+
+The master will create *N* PSs, with *N* as the number of PSs specified by the user. Each model variable and embedding vector has a corresponding PS. Thus, every PS has a subset of model variables and embedding tables.
+
+The master will monitor PS pods status similar to what it does for worker pods. In case a PS pod fails, the master will try to relaunch it. Since PS pods have higher priority than worker pods, if there are still some running worker pods, the relaunch will succeed by using either idle or preempted Kubernetes resources. If the relaunch fails, there are no worker pods left. The Elastic job has to wait for resources for the PS pod and worker pods.
+
+Each worker has a local copy of the model variables. After the master relaunches a PS, the PS pod can recover model variables from workers. For embedding vectors, PS must create replicas to support fault tolerance. For each PS *P(i)*, it will store *M* replicas in the following *M* PSs from *P(i+1 % N)* to *P(i+M % N)*. The relaunched PS can recover embedding vectors from one of its replicas. If there are more than *M* continuously-indexed PSs failing, at least one PS fails with all of its replica PSs. The ElasticDL job has to recover from a recent checkpoint.
+
+
+
+## PServer
+
+
+### Compoments
+
+
+![pserver](./images/pserver.png)
+
+
+PServer contains three main compoments:
+
+- KVStore
+- Gradient Queue
+- Optimizer
+
+1. The worker initializes a model, and pushes parameters to KVStore.
+2. Before each step of training, the worker pulls the latest model from KVStore.
+3. After a round of forward/backward computation, the worker pushes gradients to the Gradient Queue waiting for processing.
+4. Optimizer gets a gradient from the Gradient Queue.
+5. Then, Optimizer looks up the corresponding parameter from KVStore.
+6. Optimizer applies gradients to parameters, and updates parameter back to KVStore.
+
+
+### Tensor Data Structure
+
+To support data communication between pods, we introduce a `Tensor` proto message:
+
+```proto
+message Tensor {
+    enum DataType {
+        BOOL = 0;
+        INT16 = 1;
+        INT32 = 2;
+        INT64 = 3;
+        FP16 = 4;
+        FP32 = 5;
+        FP64 = 6;
+    }
+    string name = 1;
+    DataType data_type = 2;
+    repeated int64 dims = 3;
+    bytes content = 4;
+    repeated int64 indices = 5;
+}
```

Comment:
Question 1, Tensor proto definition:
-  set name to Tensor, and use `repeated Tensor param` 
- or use `map<string, Tensor> param`


I prefer to set `name` to Tensor.  Map is designed for random access. If we do not have such demand, vector is enough.

Reply:
Agree.